Add per color counter data for direct meters.

kishanps opened the pull request.

kishanps commented:
@jafingerhut @antoninbas This is a first draft of the p4runtime proto changes we discussed [here](https://github.com/p4lang/p4c/issues/2983)

smolkaj commented:
Feedback from @antoninbas: Also add counters for indirect meters.

jfingerh commented:
Just a quick note: If your goal is to enhance bmv2 and the P4Runtime API so that it more closely matches a particular model of switch ASIC, then I have no objections.  I mean, you can completely rework bmv2 and P4Runtime API so it matches exactly what switch vendor X provides, and has anywhere from minor to big differences with any other switch ASIC in the world.  I am not saying that these changes alone are going to push you into huge differences.  I am saying that if your goal is cross-switch compatibility, then tweaks like this are _potentially_ nudging you further from that goal, not closer to it.  If you want to be sure, you should be checking these kinds of changes with other switch ASICs you may be interested in.  If you ignore that, then you could reach a place where you have to significantly rework things for compatibility later (if that is part of the goal -- it may not be).

smolkaj commented:
Thanks for raising these important points, @jfingerh.

With PINS, we explicitly want to build a switch stack that is **not** tied to a particular vendor. So we definitely want to avoid "overfitting" to a particular ASIC accidentally.

I cannot speak to the capabilities of various ASICs, since that is not my expertise. FWIW, the counters we're modeling here are supported by SAI. I take that as an indication that this capability is likely supported by many vendors. This may not be a sound inference? If someone knows more on this subject, please do share.

Also, note that our proposal (not part of this PR yet, to be added) would be that supporting these counters is optional.
So from a P4Runtime perspective at least, there should hopefully be no concerns that this change will make it harder to build P4Runtime standard-compliant targets.

This raises a deeper, more general question though (that you may have been alluding to, and that I have been pondering about): how open should we be to adding features to P4Runtime for which support is optional?
At one extreme, you could imagine to disallow optional features altogether, and make P4Runtime the intersection of features supported by all vendors.  
At the other extreme, you could imagine to liberally allow optional features, and make P4Runtime the union of features supported by all vendors.

The latter extreme makes it really easy to extend the standard and to build standard compliant switches, but also make the standard almost useless because it is so liberal that it makes barely any guarantees. The former extreme is more desirable from a cross-vendor compatibility & testing perspective, but in its extreme form would be so limiting as to barely support any features. Navigating these two extremes is really tricky. Are there any guiding principles that can help us with this, or that maybe the WG has relied on in the past?

kishanps commented:
Echo @smolkaj comments, this particular counter is supported by [SAI](https://github.com/opencomputeproject/SAI/blob/05648bc62fbdf9c456c0b8d93d6ee53b5edd5df9/inc/saipolicer.h#L228) and we would expect a SAI compliant vendor to have this in their ASIC.

jafingerhut commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -4062,6 +4105,27 @@ supported as follows:
+
+Packets are marked as GREEN, YELLOW or RED based on the observed rate and the
+`MeterCounterData` helps to provide the detailed view of packets/bytes
+matching each color action that was taken.
```
And I would point out that with the suggested phrase "Whenever a meter marks a packet with a color", some people associate the verb "mark" in this context to mean actually changing something in the packet, e.g. a DSCP header field value.  Updating a P4 meter in the v1model, PSA, and TNA architectures by itself, without any other P4 code after that, does nothing to change the packet in any way, only to update the state of the meter (e.g. its token bucket state).

Maybe instead an even more limited phrase like "Whenever a meter is updated and returns a color" ?

kishanps replied:
Yep, removed the first paragraph.

kishanps commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3981,6 +4010,7 @@ direct meter entry can no longer be accessed.
 message DirectMeterEntry {
```
Moved the earlier desciption in P4Info section to here.

kishanps commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -4009,7 +4046,8 @@ A client may use `ReadRequest` in two ways to read a `DirectMeter` config.
   in the `ReadRequest`. The `table_entry.match` field must match the
   `TableEntry` whose meter config is being read. If no such entry is found, the
-  server returns the error code `NOT_FOUND`.
+  server returns the error code `NOT_FOUND`. Read responses also include the
```
Not specifying makes it consistent with the other field meter_config here. Otherwise it looks odd to have one field required and the other not. Wdyt?

kishanps replied:
Updated to reflect that the field will be populated only if present in the request message.

kishanps commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -4062,6 +4105,27 @@ supported as follows:
+The `MeterCounterData` P4Runtime message represents the per color counters
+associated with the action taken using the meter configuration. As seen above,
+they can be associated with a `DirectMeterEntry` or `MeterEntry`. However,
+these counters are implementation specific and will have the fields unset in
```
I think this ties closely with the comment in the previous section on whether the field should be set or unset in the read request.

kishanps replied:
Updated to return UNIMPLEMENTED if not supported by the target.

kishanps commented on proto/p4/v1/p4runtime.proto:
```diff
@@ -332,6 +336,7 @@ message MeterEntry {
   uint32 meter_id = 1;
   Index index = 2;
   MeterConfig config = 3;
+  MeterCounterData meter_counter_data = 4;
```
Just wanted to differentiate it between the overall counter_data and the per-color meter counters.

kishanps commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -1684,6 +1684,13 @@ type](#sec-user-defined-types). This is useful for
+of these counters are the same as the `counter_data` field. Note that certain
+implementations might not support these counters and it is okay to return an
+UNIMPLEMENTED error by the server in those circumstances.
+
```
Moved.

kishanps commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -2848,6 +2855,10 @@ the `TableEntry` entity, which has the following fields:
 
+* `meter_counter_data`, which is used to read and write the color counter values
+  for the direct meter entry attached to this table entry, if any. See [Direct
+  resources](#sec-direct-resources) section for more information.
```
Have it mentioned in the later section.

kishanps commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3318,11 +3329,11 @@ error code.
+`meter_counter_data` fields for read and write requests, based on whether the
+fields are set or not. We do not cover error cases in the list, &ie; we assume
+that we are dealing with a table which is assigned a direct counter / a direct
+meter, and that the action being used for the table entry "executes" the direct resource appropriately.
```
Given that we have mentioned about UNIMPLEMENTED later, is it still required here ?

kishanps replied:
Updated.

kishanps commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3359,6 +3370,24 @@ for the table entry "executes" the direct resource appropriately.
         * if **set**: The response includes the counter entry's value read from
           the target.
 
+* `meter_counter_data` field
```
To be consistent with the rest of the fields here, have it take effect only when the individual subfields are set ?

antoninbas replied:
I feel like @smolkaj has a good point here. Would it make sense to say that if `meter_counter_data` is set, all subfields must also be set?

kishanps replied:
The text below this line mentions about the default configuration which means that all packets are marked green, so would it be ok to have the read request specify other colors ?

antoninbas replied:
> The text below this line mentions about the default configuration which means that all packets are marked green, so would it be ok to have the read request specify other colors ?

This is about meter configuration, so I am not sure how this relates to writing / reading the meter's counters?

To clarify @smolkaj's question, consider the case (WriteRequest) where the `meter_counter_data` field is set but not all nested fields are set, e.g. `green` is not set, `yellow` is set to 0 and `red` is not set. What is the expected behavior? One reasonable behavior would be resetting the yellow counter to 0, while leaving the other 2 counters untouched. However, maybe some targets only support resetting all counters simultaneously.

Another possible option:
    * `WriteRequest (INSERT)`
        - if **unset**: The initial value for all 3 counter entries is the default (0).
        - if **set**: Same as **unset** case, subfields are ignored.
    * `WriteRequest (MODIFY)`
        - if **unset**: The 3 counter entry values are not changed.
        - if **set**: All 3 counter entries are reset to 0, subfields are ignored.
    * `ReadRequest`
        * if **unset**: The response does not include the 3 counter entry values.
        * if **set**: The response includes all 3 counter values read from the target.

With this approach, I think we keep things simple. Subfields being set / unset in requests is unrelevant.

The bottom line is that the current version here is to vague because it ignores the fact that there *are* subfields (unlike for normal direct counters).

kishanps replied:
The closes precedence I see is non-key fields being ignored for a delete operation of a table entry.  Will let @antoninbas chip in.

kishanps replied:
@antoninbas I am leaning towards making the INSERT & MODIFY return INVALID_ARGUMENT for non-zero values. That way, we dont change anything and the behavior is well defined. Wdyt ?

antoninbas replied:
Sounds fine to me.

kishanps replied:
Updated.

kishanps commented:
@jfingerh, @antoninbas Gentle reminder.

antoninbas commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -1684,6 +1684,13 @@ type](#sec-user-defined-types). This is useful for
+the per color meter counter data. It has a granular list of the counters
+applicable to the possible color actions such as RED, YELLOW and GREEN.The unit
+of these counters are the same as the `counter_data` field. Note that certain
+implementations might not support these counters and it is okay to return an
+UNIMPLEMENTED error by the server in those circumstances.
```
I feel like this should be justified more, probably by mentioning that the primary purpose of these counters is for debugging / troubleshooting.

antoninbas reviewed (commented):
sorry for the delay and thanks for the ping, holidays got in the way

kishanps commented:
Addressed review comments.

antoninbas commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -4062,6 +4107,29 @@ supported as follows:
+
+~ Begin Proto
+message MeterCounterData {
+  CounterData green_counter_data = 1;
+  CounterData yellow_counter_data = 2;
+  CounterData red_counter_data = 3;
```
should be updated to reflect new proto field names

kishanps commented:
Resolved open comments, PTAL.

antoninbas commented:
@kishanps CI is currently failing:
```
The generated Go files are not up-to-date
You can regenerate them with './codegen/update_go.sh' and commit the changes
```

antoninbas commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3359,6 +3363,21 @@ for the table entry "executes" the direct resource appropriately.
+        * if **unset**: The initial value for all 3 counter entries is the default
+          (0).
+        * if **set**: The initial value for all 3 counter entries is the default (0).
+        Sub-field values, if any set, are ignored. 
```
nit: inconsistent indentation with the rest

kishanps replied:
I believed indentation will be taken care of by the Madoko tool.

antoninbas commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3359,6 +3363,21 @@ for the table entry "executes" the direct resource appropriately.
+        * if **set**: The initial value for all 3 counter entries is the default (0).
+        Sub-field values, if any set, are ignored. 
+    * `WriteRequest (MODIFY)`
+        * if **unset**: All the 3 counter entry's value is not changed.
+        * if **set**: All the 3 counter entry's value is reset to 0. 
```
not grammatically correct. For `set`, I believe we should emphasize that subfields are ignored, just like for `INSERT`.

antoninbas commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3359,6 +3363,21 @@ for the table entry "executes" the direct resource appropriately.
+        * if **unset**: All the 3 counter entry's value is not changed.
+        * if **set**: All the 3 counter entry's value is reset to 0. 
+    * `ReadRequest`
+        * if **unset**: The response does not include the counter entry's value
```
> does not include the the counter values

antoninbas commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3359,6 +3363,21 @@ for the table entry "executes" the direct resource appropriately.
+    * `ReadRequest`
+        * if **unset**: The response does not include the counter entry's value
+          (`meter_counter_data` is unset in the response).
+        * if **set**: The response includes all the 3 counter entry's value read from
+          the target.
```
same comment as above, grammatically incorrect. Maybe: "The response includes all 3 counter values read from the target"

antoninbas commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -4062,6 +4106,28 @@ supported as follows:
+a color, the corresponding color counter GREEN, YELLOW or RED is updated. 
+
+As seen above, these counters can be associated with a `DirectMeterEntry` or `MeterEntry`.
+However, they are implementation specific and targets not capable of supporting these counters should return `UNIMPLEMENTED` if the field(s) was set in the initial read or write request.
```
wrap line

kishanps commented:
Thanks @antoninbas for pointing out the script, builds passing now.

kishanps commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3955,6 +3976,14 @@ stream using two configured rates --- the Peak Information Rate (PIR) and
 
+`Meter` & `DirectMeter` have an additional field `counter_data` that holds
+the per color meter counter data. It has a granular list of the counters
+applicable to the possible meter color such as GREEN, YELLOW and RED. The unit
```
Added.

kishanps commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3996,9 +4026,18 @@ A `WriteRequest` may only include an `Update` message of type `MODIFY` with a
-`INVALID_ARGUMENT` in that case.
+`INVALID_ARGUMENT` in that case. Targets not supporting per color meter
+counters can reject a write request that has this field set with an
+`UNIMPLEMENTED` error code.
```
The earlier paragraph is for MODIFY and this note is for INSERT or DELETE.

smolkaj replied:
This does not make sense to me. If you have an update of type `INSERT` or `DELETE` such that `update.entity().direct_meter_entry().has_counter_data()` is true, then this implies that `update.entity().has_direct_meter_entry()` is true, which means we must return `INVALID_ARGUMENT` by the first sentence in this paragraph.

smolkaj replied:
> can reject a write request that has this field set with an `UNIMPLEMENTED` error code.

Can we make this unambiguous, "can reject" -> "must reject"?

On a related note, maybe it would be better to state a prior that any `DirectMeterEntry` message with `counter_data` set must be rejected with UNIMPLEMENTED by implementations that do not support it. That way we don't have to repeat this again and again for the various types of requests, and there is no ambiguity.

kishanps replied:
Updated the unimplemented part in the 'counter_data' field text and removed the repetitions.

kishanps commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3996,9 +4026,18 @@ A `WriteRequest` may only include an `Update` message of type `MODIFY` with a
   specified by the client. Note that if this Protobuf field is not set, the
   meter config is set to execute the default behavior (GREEN for all packets).
 
+* `counter_data` is used to set the per color counter values for the
```
Updated.

kishanps commented:
Gentle reminder.

smolkaj commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3955,6 +3976,14 @@ stream using two configured rates --- the Peak Information Rate (PIR) and
 
+`Meter` & `DirectMeter` have an additional field `counter_data` that may hold
+per color counter data. It has a granular list of the counters applicable to the
+possible meter colors such as GREEN, YELLOW and RED. The counter associated with
```
Nit: remove "such as", since GREEN, YELLOW, RED is the exact list of colors we support.

smolkaj commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3955,6 +3976,14 @@ stream using two configured rates --- the Peak Information Rate (PIR) and
 Committed Information Rate (CIR), and their associated burst sizes --- and
 "marks" its packets as GREEN, YELLOW or RED based on the observed rate.
 
+`Meter` & `DirectMeter` have an additional field `counter_data` that may hold
```
`Meter` -> `MeterEntry`
`DirectMeter` -> `DirectMeterEntry`

smolkaj commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3955,6 +3976,14 @@ stream using two configured rates --- the Peak Information Rate (PIR) and
+a specific color is incremented when a packet is "marked" with that color. The
+primary purpose of the color counters is for debugging purposes. Note that
+certain implementations might not support these counters and it is okay to
+return an `UNIMPLEMENTED` error by the server in those circumstances.
```
> "in those circumstance"

In what circumstance? We need to make this unambiguous.

smolkaj replied:
I think with the changes I suggested above, this could simply be removed.

smolkaj commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3955,6 +3976,14 @@ stream using two configured rates --- the Peak Information Rate (PIR) and
 "marks" its packets as GREEN, YELLOW or RED based on the observed rate.
 
+`Meter` & `DirectMeter` have an additional field `counter_data` that may hold
+per color counter data. It has a granular list of the counters applicable to the
```
"that may hold per color counter data for targets that support it, and that must always be unset for targets that do not support it."

smolkaj commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -4062,6 +4109,26 @@ supported as follows:
+### `MeterCounterData`
+
+The `MeterCounterData` P4Runtime message represents the per color counters
+associated with the meter configuration. Whenever a meter is updated and returns
```
"with the meter configuration" -> "with a meter" (or "with a meter entry" if you prefer).

smolkaj replied:
Nit: 
> "Whenever a meter is updated"

Can we say "executed" instead, to match the terminology used in  psa.p4>

smolkaj commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -4062,6 +4109,26 @@ supported as follows:
+a color, the corresponding color counter GREEN, YELLOW or RED is updated.
+
+As seen above, these counters can be associated with a `DirectMeterEntry` or
+`MeterEntry`. However, they are implementation specific and targets not capable
```
I don't think these counters are "implementation specific". They can either be supported, or not supported, and in both cases this standard defines how they should behave.

kishanps replied:
Removed the implementation specific

smolkaj commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -4062,6 +4109,26 @@ supported as follows:
+
+As seen above, these counters can be associated with a `DirectMeterEntry` or
+`MeterEntry`. However, they are implementation specific and targets not capable
+of supporting these counters should return `UNIMPLEMENTED` if the field(s) was
```
> if the field(s) was set

Can we just say "if a `MeterCounterData` field is set`? This is more precise and unambiguous.

smolkaj reviewed (changes requested):
Final nits.

The remaining big question is whether returning `UNIMPLEMENTED` is acceptable (technically a breaking change) or not. @antoninbas @jfingerh

jfingerh commented:
"The remaining big question is whether returning UNIMPLEMENTED is acceptable (technically a breaking change) or not."

So here is a question I love to ask to people who claim to advocate semantic versioning, to find out how much they actually believe it: Are you willing to bump up the major version number of the P4Runtime spec to 2.0.0 in the next release if this change is included?

If yes, then I have no objections to backwards-incompatible changes at all.  Having a brief note in the change log for the revisions explaining what the known backwards-incompatible changes are would be a nice reference, for those deveopers who want a summary of what those backwards-incompatible changes are.

If no, then why not?

antoninbas commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3318,11 +3322,12 @@ error code.
 We leverage Protobuf's ability to differentiate between set and unset fields to
```
The spec should be backward compatible. However it's unclear to me why returning UNIMPLEMENTED would break backwards compatibility here, so maybe you could clarify with an example? In particular, as far as older clients are concerned, nothing would change. For newer clients, yes the behavior would be different based on whether the server is old (fields ignored) or new but doesn't support the feature (UNIMPLEMENTED returned). But I wouldn't say that it breaks backward compatibility. We have been using this document (https://cloud.google.com/apis/design/compatibility#adding_a_field_to_a_request_message) as a reference IIRC, and all I see there is:

> Adding request fields can be non-breaking, so long as clients that don't specify the field will be treated the same in the new version as in the old version.

TLDR; my opinion here is that because old implementations are not impacted (e.g. an old client), we are not breaking backwards compatibility. New clients may need to handle different behaviors (based on whether the server is old or new), but these are new clients so it should be ok.

kishanps replied:
I agree.
In spirit of that, is it necessary to mention that the counters will be untouched for (unset) INSERT requests ? Can we say the default value is left to the target implementation and is generally 0.
>WriteRequest (INSERT)`
>        * if **unset**: The initial value for all 3 counter entries is the default (0).

smolkaj replied:
Thanks for explaining @antoninbas, that makes sense.

> is it necessary to mention that the counters will be untouched for (unset) INSERT requests ? 

A direct meter associates meter state with each table entry. This state does not exist / is not defined before the table entry gets created, as far as v1model or PSA are concerned. So what does it even mean to leave the counters (which are part of the meter state for the table entry) "untouched"? I don't think there is a sensible definition of "untouched"?

> Can we say the default value is left to the target implementation and is generally 0.

Since we already ask targets to support resetting counters to zero, why not do that by default when creating a new entry?

kishanps replied:
I was highlighting this from a compatibility point of view. If targets were setting the meter color counters to a non-zero value (although very unlikely) before, on an INSERT operation, targets complying to the updated spec now need to set it to zero.

It is an unlikely scenario, so I don't mind ignoring this corner case scenario for backward compatibility.

smolkaj commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -4062,6 +4111,26 @@ supported as follows:
+As seen above, these counters can be associated with a `DirectMeterEntry` or
+`MeterEntry`. Targets not capable of supporting these counters should return
+`UNIMPLEMENTED` if a `MeterCounterData` field was set in the initial read or
+write request.
```
"in the initial read or write request." -> "in a read or write request"?

smolkaj commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3996,9 +4028,18 @@ A `WriteRequest` may only include an `Update` message of type `MODIFY` with a
+* `counter_data` is used to set the per color counter values to the default
+  value of (0), which is essentially clearing the counters. If the field is
+  unset, the counter values are left untouched. If the field is set, targets
+  supporting this counter should reset all the color counters to (0),
```
> targets supporting this counter should reset all the color counters to (0).

It depends. If any of the subfields are nonzero, targets supporting the counter should presumably return INVALID_ARGUMENT, to be consistent with the spec for TableEntry?

kishanps commented:
The latest builds are failing, comparing a good and failed run, looks like the environment has changed. The error seems to be happening on a timeout where an input is expected for configuring tzdata, whereas the previous success runs did not setup tzdata at all. @antoninbas Any recent changes on the build side ?

```
Setting up tzdata (2021e-0ubuntu0.20.04) ...
debconf: unable to initialize frontend: Dialog
debconf: (TERM is not set, so the dialog frontend is not usable.)
debconf: falling back to frontend: Readline
Configuring tzdata
------------------
Please select the geographic area in which you live. Subsequent configuration
questions will narrow this down by presenting a list of cities, representing
the time zones in which they are located.
  1. Africa      4. Australia  7. Atlantic  10. Pacific  13. Etc
  2. America     5. Arctic     8. Europe    11. SystemV
  3. Antarctica  6. Asia       9. Indian    12. US
Geographic area: 
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
```

smolkaj commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -3955,6 +3978,15 @@ stream using two configured rates --- the Peak Information Rate (PIR) and
 
+`MeterEntry` & `DirectMeterEntry` have an additional field `counter_data` that
+may hold per color counter data for targets that support it, and that must
+always be unset for targets that do not support it. If set and unsupported by
```
"If set" -> "If set in a request"

smolkaj reviewed (approved):
LGTM

kishanps commented:
@antoninbas I resolved the 2 proto generated files to my best knowledge but my merge commit build failed.

In my local repo, I did a git pull --rebase, bazel build under the proto dir and then codegen/update.sh. It generated a bunch of file changes (even in p4data which I didnt touch in my PR). And most of the changes seem like they were reversing the protoc changes you did in your commit last week (see below). Am I missing something ? 

```
git diff go/p4/v1/p4runtime.pb.go
diff --git a/go/p4/v1/p4runtime.pb.go b/go/p4/v1/p4runtime.pb.go
index 52f850f..6b86f4b 100644
--- a/go/p4/v1/p4runtime.pb.go
+++ b/go/p4/v1/p4runtime.pb.go
@@ -15,7 +15,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
 //     protoc-gen-go v1.26.0
-//     protoc        v3.18.1
+//     protoc        v3.6.1
 // source: p4/v1/p4runtime.proto
 
git status
	modified:   go.mod
	modified:   go/p4/config/v1/p4info.pb.go
	modified:   go/p4/config/v1/p4types.pb.go
	modified:   go/p4/v1/p4data.pb.go
	modified:   go/p4/v1/p4runtime.pb.go
	modified:   py/p4/config/v1/p4info_pb2.py
	modified:   py/p4/config/v1/p4info_pb2_grpc.py
	modified:   py/p4/config/v1/p4types_pb2.py
	modified:   py/p4/config/v1/p4types_pb2_grpc.py
	modified:   py/p4/v1/p4data_pb2.py
	modified:   py/p4/v1/p4data_pb2_grpc.py
	modified:   py/p4/v1/p4runtime_pb2.py
	modified:   py/p4/v1/p4runtime_pb2_grpc.py

```

antoninbas commented:
@kishanps maybe try doing `docker pull third-party:latest` first. I'm assuming you still have the old version of the image from before.

smolkaj commented:
Looks good from my side. @antoninbas @jafingerhut any other changes needed here?

antoninbas reviewed (approved):
LGTM

kishanps commented:
Thanks @smolkaj @antoninbas @jfingerh for all the help!

antoninbas commented:
@kishanps Thanks for contributing this!
